docker/Dockerfile

```diff
@@ -5,46 +5,104 @@ ENV LANG en_US.utf8
 RUN apt-get update \
   && apt-get -y install ca-certificates build-essential libsasl2-dev openjdk-11-jdk
 
-FROM base AS builder
+FROM base AS rust-base
 
-RUN apt-get update && apt-get -y install make cmake protobuf-compiler curl bash lld maven unzip
+RUN apt-get -y install make cmake protobuf-compiler curl bash lld unzip
 
 SHELL ["/bin/bash", "-c"]
 
 RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- --no-modify-path --default-toolchain none -y
+ENV PATH /root/.cargo/bin/:$PATH
+ENV CARGO_INCREMENTAL=0
 
-RUN mkdir -p /risingwave
+COPY rust-toolchain rust-toolchain
 
-WORKDIR /risingwave
+# We need to add the `rustfmt` dependency, otherwise `risingwave_pb` will not compile
+RUN rustup self update \
+  && rustup set profile minimal \
+  && rustup show \
+  && rustup component add rustfmt
+
+RUN cargo install --git https://github.com/xxchan/cargo-chef cargo-chef --locked --rev 11f9fed
 
+FROM rust-base AS rust-planner
+
+RUN mkdir -p /risingwave
+WORKDIR /risingwave
 COPY ./ /risingwave
 
-ENV PATH /root/.cargo/bin/:$PATH
+RUN cargo chef prepare --recipe-path recipe.json
 
-ENV IN_CONTAINER=1
+FROM rust-base AS rust-builder
 
-ARG GIT_SHA
-ENV GIT_SHA=$GIT_SHA
+RUN mkdir -p /risingwave
+WORKDIR /risingwave
 
+COPY --from=rust-planner /risingwave/recipe.json recipe.json
+
+# Build dependencies - this can be cached if the dependencies don't change
+RUN cargo chef cook --release --recipe-path recipe.json
```

Comment:
From my understanding, we depend on the cache of the `docker build` itself. However, I am concerned that it may not be effective in most cases because the `Cargo.lock` file is frequently modified (more than once per day).

Reply:
You are right. So we will need to disable/decrease frequency of dependabot to make the cache effective. 

However, I think the changes doesn't have any drawbacks (?). And it can also benefit local docker build (if anyone does it).

Reply:
Otherwise, maybe we just also use sccache in it/

Reply:
> the changes doesn't have any drawbacks (?).

Maybe it will slightly slow down compilation.

So do you prefer keep or delay `cargo chef`?